docs/fidi_lint.1
.\" // Copyright 2018-2019 Google LLC
.\"
.\" Licensed under the Apache License, Version 2.0 (the "License");
.\" you may not use this file except in compliance with the License.
.\" You may obtain a copy of the License at
.\"
.\" https://www.apache.org/licenses/LICENSE-2.0
.\"
.\" Unless required by applicable law or agreed to in writing, software
.\" distributed under the License is distributed on an "AS IS" BASIS,
.\" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\" See the License for the specific language governing permissions and
.\" limitations under the License.
.TH FIDI_LINT 1 2018-12-29
.SH NAME
fidi_lint \- A lint checker for fidi input files
.SH SYNOPSIS
.B fidi_lint
.RI "<request file>"
.br
.RI "cat <request file> |"
.B fidi_lint
.br
.RI "cat <request file> |"
.B fidi_lint \-
.SH DESCRIPTION
This manual page documents the
.B fidi_lint
syntax and sanity checker for
.B fidi (φίδι)
request files. (See
.BR fidi_request (5)
for details on the request language).  At the simplest level: this is
a simple HTTP application that talks to other instances of itself. The
number and targets of these calls are defined by the input
request.
.PP
The request describes a series of calls between HTTP servers
simulating the behaviour of a service being mocked, often starting
with a single call, and the cascade of calls that result by the first
server making more HTTP requests. Essentially, the request defines:
.IP Nodes 4
all the nodes (with a host and port, or a URL) touched in the sequence
of calls.
.IP Calls
It also defines all the calls (edges between nodes, with
source and destinations, and call attributes). In most cases this will
describe a directed graph.
.PP
\fBfidi_lint\fP checks the input request for syntax errors; and makes
a best effort attempt to recover from errors (popping current state
and discarding tokens until it re-establishes a good state) in order
to find as many errors in the input in the same check.
.PP
If the request contains no syntax errors,
.B fidi_lint
runs sanity checks (ensures that arguments that should be numeric are
indeed so, and that they are within bound, for example, and that the
host attribute list has sufficient information to succesfully make a
call). Any issues discovered are reported.
.PP
Finally,
.B fidi_lint
produces a desciption of the graph, annotating each edge with the call
sequence, in a format that can be fed to
.RI dot
to produce a visual representation of the request.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`\-').
A summary of options is included below.
.TP
.B \-h, \-\-help
Show summary of options, and exit.
.TP
.B \-v, \-\-version
Show version of program, and exit.
.TP
.B \-
An indication that
.B fidi_lint
should read the request to be checked from the standard input..
.SH EXAMPLES
Given the following input:
.PP
.RS 4
.EX
fidi_lint << EOF
client    [ hostname = "127.0.0.1", port = 8001, ]
frontend  [ hostname = "127.0.0.1", port = 8002, ]
[
  predelay = 12,
  postdelay = 20,
  response  = 200,
  -> frontend repeat = 2 sequence = 1 [
      predelay = 12,
      postdelay = 20,
      response  = 200,
    ]
]
EOF
.EE
.RE
.PP
.B fidi_lint
produces the output
.PP
.RS
.EX
digraph fidi {
  node [shape=record];
  client [ label="{hostname=127.0.0.1|port=8001|client}" ];
  frontend [ label="{hostname=127.0.0.1|port=8002|frontend}" ];

  Source -> TopNode [ label="1"]
    // postdelay = 20,
    // predelay = 12,
    // response = 200,

  TopNode -> frontend [ label="1.1"]
    // postdelay = 20,
    // predelay = 12,
    // response = 200,
}
.EE
.RE
.PP
The output can then be fed to
.B dot(1)
to produce, say, a
.RI .png
format graph.
.SH "SEE ALSO"
.BR fidi_app (1),
.BR fidi_request (5),
.BR dot (1).
.SH BUGS
None known so far.
.SH AUTHOR
Manoj Srivastava <srivasta@google.com>


NEWS
fidi (φίδι) NEWS                                               -*-indented-text-*-
  History of user-visible changes.
  3 January 2019

All changes mentioned here are more fully described in the fidi (φίδι)
manual, which is contained in this distribution as Doxygen
documentation.  See the README file and the fidi (φίδι) manual for
instructions for reporting bugs.

Version 1.0.0 (10 Jan 2019)

* First version

Copyright 2018-2019 Google LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


src/fidi_lint.cc
// -*- mode: c++; -*-
// fidi_lint.cc ---

// Copyright 2018-2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// \file
/// \ingroup lint
///
/// This is the main file of the fidi (φίδι) lint tool. Here we
/// implement commands line parsing, and then either put out a help
/// screen or create a parser and check the input.

// Code:

#include <sys/stat.h>
#include <sys/types.h>

#include <unistd.h>
#include <cstdlib>
#include <cstring>
#include <iostream>

#include "src/fidi_lint_driver.h"

/// \brief  Main function
///
/// \details Implement command line parsing. --help or --version are
/// terminal options, in the sense that we print out the help text or
/// version number and then exit. Otherwise, first, create a parser.
/// Secondly, determine if a filename has been provided, or we should
/// read from ther standard input; and then pass the appropriate input
/// stream to the parser, and then perform the syntax and sanity
/// checks.  Finally, it generates a graph description in dot(1)
/// format, which can then be processed to create a graph of the
/// request and the cascade of the resulting requests that it
/// generates, to provide a visual depiction of the requested
/// behaviour.
///
/// \param[in]  argc number of arguments
/// \param[in]  argv An array of character pointers containing the arguments
///
/// \return an integer 0 upon exit success
int
main(const int argc, const char **argv) {
  fidi::LintDriver driver;

  try {
    if (argc == 1) {
      driver.Parse(std::cin);
    } else if (argc == 2) {
      /* simple help menu */
      if ((std::strncmp(argv[1], "-h", 2) == 0) ||
          (std::strncmp(argv[1], "--h", 3) == 0)) {
        std::cout << PACKAGE_NAME << " lint usage\n\n"
                  << "use cat to  pipe a file to std::cin\n"
                  << "or give a filename to validate a file\n"
                  << "    fidi_lint input.txt\n"
                  << "    cat input.txt | fifi_lint\n\n"
                  << "Use -v or --version to get the version\n"
                  << "    fidi_lint -v\n"
                  << "    fidi_lint --version\n\n"
                  << "use -h or --help to get this menu\n";
        return (EXIT_SUCCESS);
      }

      if ((std::strncmp(argv[1], "-v", 2) == 0) ||
          (std::strncmp(argv[1], "--v", 3) == 0)) {
        std::cout << PACKAGE_NAME << " version " << PACKAGE_VERSION << "\n";
        return (EXIT_SUCCESS);
      }

      if (std::strncmp(argv[1], "-", 1) == 0) {
        // Support - as synonym for file stdin
        driver.Parse(std::cin);
      } else {
        struct stat sb;
        if (lstat(argv[1], &sb) == -1) {
          std::cerr << "Unknown file or option: " << argv[1] << std::endl
                    << "Usage\n"
                    << "    fidi_lint input.txt\n"
                    << "    cat input.txt | fifi_lint\n"
                    << "use cat to  pipe to the standard input.\n"
                    << "just give a filename to validate a file\n"
                    << "use -h to get this menu\n";
          return (EXIT_FAILURE);
        }
        driver.Parse(argv[1]);
      }
    } else {
      std::cerr << "Unknown arguments. We expect 0 or 1.\n"
                << "Usage\n"
                << "    fidi_lint input.txt\n"
                << "    cat input.txt | fifi_lint\n"
                << "use cat to  pipe to the standard input.\n"
                << "just give a filename to validate a file\n"
                << "use -h to get this menu\n";
      return (EXIT_FAILURE);
    }
  } catch (std::bad_alloc &ba) {
    std::cerr << "Failed to allocate scanner: (" << ba.what() << ")\n";
    return (EXIT_FAILURE);
  }

  // At this point, we have run the parser
  if (driver.nerrors_ != 0) {
    std::cerr << "Proceeding despite failures. "
              << "The graph is likely inaccurate." << std::endl;
  }
  driver.Execute(std::cout) << std::endl;
  if (driver.nerrors_ != 0) { return (EXIT_FAILURE); }

  return (EXIT_SUCCESS);
}

//
// fidi_lint.cc ends here
